### PR TITLE
[Velox To Substrait] Add support for IF and SWITCH expressions

### PR DESCRIPTION
**Pull request by chaojun-zhang**

### CODE REVIEW

File: velox/substrait/VeloxToSubstraitExpr.cpp [hunk 1]
```diff
@@ -158,7 +156,24 @@ const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
         typeConvertor_->toSubstraitType(arena, callTypeExpr->type()));
 
   } else {
-    VELOX_NYI("Unsupported function name '{}'", functionName);
+    if (callTypeExpr->inputs().size() % 2 != 1) {
+      VELOX_NYI(
+          "Number of arguments are always going to be odd for if/then or switch expression");
```
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitExpr.cpp:161]: Is this assumption accurate? Switch statement has an even number of input when "else" branch is not specified (in this case it is assumed "else null").

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 2]
```diff
@@ -124,10 +124,41 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+    params.emplace_back(ifClauseExpr);
+    auto thenClauseExpr = toVeloxExpr(ifExpr.then(), inputType);
+    params.emplace_back(thenClauseExpr);
+
+    if (!thenClauseExpr->type()->containsUnknown()) {
+      resultType = thenClauseExpr;
+    }
+  }
+
+  VELOX_CHECK(
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:153]: VELOX_CHECK_NOT_NULL

**Review by mbasmanova — COMMENTED** [review]
@chaojun-zhang Some comments.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 1 above)
Comment (chaojun-zhang) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: updated

File: velox/substrait/SubstraitToVeloxExpr.cpp (hunk 2 above)
Comment (chaojun-zhang) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Fixed

Comment (chaojun-zhang) [reply velox/substrait/SubstraitToVeloxExpr.cpp]: Fixed

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 1 above)
Comment (chaojun-zhang) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: Updated

File: velox/substrait/VeloxToSubstraitExpr.cpp [hunk 3]
```diff
@@ -158,7 +156,25 @@ const ::substrait::Expression& VeloxToSubstraitExprConvertor::toSubstraitExpr(
...
   } else {
-    VELOX_NYI("Unsupported function name '{}'", functionName);
+    // For today's version of Substrait, you'd have to use IfThen if you need
+    // the switch cases to be call typed expression.
+
+    size_t inputsSize = callTypeExpr->inputs().size();
+    bool hasElseInput = inputsSize % 2 == 1;
+
+    auto ifThenExpr = substraitExpr->mutable_if_then();
+    size_t last = hasElseInput ? inputsSize - 1 : inputsSize;
```
Comment (mbasmanova) [inline velox/substrait/VeloxToSubstraitExpr.cpp:166]: This variable is not needed. You can use inputsSize instead.

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 4]
```diff
@@ -124,10 +124,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
       VELOX_NYI(
           "Substrait conversion not supported for Expression '{}'", typeCase);
   }
 }
 
+std::shared_ptr<const core::ITypedExpr>
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression_IfThen& substraitIfThen,
+    const RowTypePtr& inputType) {
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:139]: TypedExprPtr

File: velox/substrait/SubstraitToVeloxExpr.cpp [hunk 5]
```diff
@@ -124,10 +124,50 @@ SubstraitVeloxExprConverter::toVeloxExpr(
...
+  }
+
+  if (substraitIfThen.has_else_()) {
+    auto elseClauseExpr = toVeloxExpr(substraitIfThen.else_(), inputType);
+    params.emplace_back(elseClauseExpr);
+    if (!resultType && !elseClauseExpr->type()->containsUnknown()) {
+      resultType = elseClauseExpr;
+    }
+  } else {
+    if (resultType) {
```
Comment (mbasmanova) [inline velox/substrait/SubstraitToVeloxExpr.cpp:160]: This branch is not needed. It is ok to omit "null" else clause.

**Review by mbasmanova — COMMENTED** [review]
@chaojun-zhang Looks good % a couple of nits.

**Review by mbasmanova — APPROVED** [review]
@chaojun-zhang Looks good to me. Thank you for the contribution.

File: velox/substrait/VeloxToSubstraitExpr.cpp (hunk 3 above)
Comment (chaojun-zhang) [reply velox/substrait/VeloxToSubstraitExpr.cpp]: updated

**Pull request closed without merging**